File: Sources/DebridStreamer/Services/Indexers/TorznabIndexer.swift
import Foundation

actor TorznabIndexer: TorrentIndexer {
    let name: String
    private let baseURL: String
    private let endpointPath: String
    private let apiKey: String?
    private let categoryFilter: String?
    private let sendAPIKeyAsHeader: Bool
    private let session: URLSession

    init(
        name: String,
        baseURL: String,
        endpointPath: String,
        apiKey: String?,
        categoryFilter: String? = nil,
        sendAPIKeyAsHeader: Bool = false,
        session: URLSession = AppHTTP.api
    ) {
        self.name = name
        self.baseURL = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        self.endpointPath = endpointPath.trimmingCharacters(in: .whitespacesAndNewlines)
        self.apiKey = apiKey
        self.categoryFilter = categoryFilter
        self.sendAPIKeyAsHeader = sendAPIKeyAsHeader
        self.session = session
    }

    func search(imdbId: String, type: MediaType, season: Int?, episode: Int?) async throws -> [TorrentResult] {
        return try await executeWithFallbacks(imdbId: imdbId, season: season, episode: episode)
    }

    func searchByQuery(query: String, type: MediaType) async throws -> [TorrentResult] {
        let trimmedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedQuery.isEmpty { return [] }

        let params: [String: String] = [
            "t": "search",
            "q": trimmedQuery,
        ]
        return try await execute(params: params)
    }

    private func executeWithFallbacks(
        imdbId: String,
        season: Int?,
        episode: Int?
    ) async throws -> [TorrentResult] {
        let trimmedImdbId = imdbId.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedIDs = buildFallbackIMDbIDs(trimmedImdbId)

        var attempts: [[String: String]] = []
        if !trimmedImdbId.isEmpty {
            for imdb in normalizedIDs {
                var withEpisode: [String: String] = ["t": "search", "imdbid": imdb]
                if let season {
                    withEpisode["season"] = String(season)
                }
                if let episode {
                    withEpisode["ep"] = String(episode)
                }

                var withQuery: [String: String] = ["t": "search", "q": imdb]
                if let season {
                    withQuery["season"] = String(season)
                }
                if let episode {
                    withQuery["ep"] = String(episode)
                }

                attempts.append(withEpisode)
                attempts.append(withQuery)
            }

            if season != nil || episode != nil {
                for imdb in normalizedIDs {
                    attempts.append(["t": "search", "imdbid": imdb])
                    attempts.append(["t": "search", "q": imdb])
                }
            }
        }

        var lastError: Error?
        var anyAttemptSucceeded = false

        for params in attempts {
            do {
                let items = try await execute(params: params)
                anyAttemptSucceeded = true
                if !items.isEmpty {
                    return items
                }
            } catch {
                lastError = error
            }
        }

        if anyAttemptSucceeded {
            return []
        }
        if let lastError {
            throw lastError
        }

        return []
    }

    private func buildFallbackIMDbIDs(_ imdbId: String) -> [String] {
        guard !imdbId.isEmpty else { return [] }

        var ids: [String] = [imdbId]
        let hasTTPrefix = imdbId.lowercased().hasPrefix("tt")
        if hasTTPrefix, imdbId.count > 2 {
            let numeric = String(imdbId.dropFirst(2))
            ids.append(numeric)
        }

        var seen = Set<String>()
        return ids.filter { id in
            guard !seen.contains(id) else { return false }
            seen.insert(id)
            return true
        }
    }

    private func execute(params: [String: String]) async throws -> [TorrentResult] {
        let request = try makeRequest(params: params)
        let (data, response) = try await session.data(for: request)
        // Throw on non-2xx (consistent with the built-in indexers) so that a
        // misconfigured endpoint / bad key surfaces as a recorded indexer failure
        // instead of an indistinguishable empty result. IndexerManager.searchAll
        // catches per-indexer, so this never aborts sibling indexers.
        guard let http = response as? HTTPURLResponse,
              (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }

        let items = try TorznabFeedParser.parse(data: data)
        return items.compactMap { item in
            let hash = item.infoHash ?? extractInfoHash(from: item.magnetURL)
            guard let hash, !hash.isEmpty else { return nil }
            return TorrentResult.fromSearch(
                infoHash: hash,
                title: item.title,
                sizeBytes: item.size,
                seeders: item.seeders,
                leechers: item.peers,
                indexerName: name,
                magnetURI: item.magnetURL
            )
        }
    }

    private func makeRequest(params: [String: String]) throws -> URLRequest {
        guard var baseComponents = URLComponents(string: baseURL) else {
            throw URLError(.badURL)
        }
        if !endpointPath.isEmpty {
            baseComponents.path = normalizedPath(baseComponents.path, endpointPath)
        }

        var queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
        if let categoryFilter, !categoryFilter.isEmpty {
            queryItems.append(URLQueryItem(name: "cat", value: categoryFilter))
        }
        if let apiKey, !apiKey.isEmpty, !sendAPIKeyAsHeader {
            queryItems.append(URLQueryItem(name: "apikey", value: apiKey))
        }
        baseComponents.queryItems = queryItems

        guard let url = baseComponents.url else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        if let apiKey, !apiKey.isEmpty, sendAPIKeyAsHeader {
            request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        }
        return request
    }

    private func normalizedPath(_ currentPath: String, _ endpoint: String) -> String {
        let current = currentPath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let append = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if current.isEmpty {
            return "/\(append)"
        }
        if append.isEmpty {
            return "/\(current)"
        }
        return "/\(current)/\(append)"
    }

    private func extractInfoHash(from magnetURL: String?) -> String? {
        guard let magnetURL, let components = URLComponents(string: magnetURL) else { return nil }
        guard let xt = components.queryItems?.first(where: { $0.name.lowercased() == "xt" })?.value else {
            return nil
        }
        let prefix = "urn:btih:"
        guard xt.lowercased().hasPrefix(prefix) else { return nil }
        return String(xt.dropFirst(prefix.count))
    }
}

private struct TorznabFeedItem {
    var title: String
    var magnetURL: String?
    var infoHash: String?
    var size: Int64
    var seeders: Int
    var peers: Int
}

private enum TorznabFeedParser {
    static func parse(data: Data) throws -> [TorznabFeedItem] {
        let delegate = TorznabXMLDelegate()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        guard parser.parse() else {
            throw parser.parserError ?? URLError(.cannotParseResponse)
        }
        return delegate.items
    }
}

private final class TorznabXMLDelegate: NSObject, XMLParserDelegate {
    private(set) var items: [TorznabFeedItem] = []

    private var currentItem: CurrentItem?
    private var currentText = ""

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        let element = elementName.lowercased()
        currentText = ""

        if element == "item" {
            currentItem = CurrentItem()
            return
        }

        guard currentItem != nil else { return }

        if element == "enclosure", let url = attributeDict["url"] {
            currentItem?.magnetURL = url
        } else if element == "torznab:attr" {
            let name = attributeDict["name"]?.lowercased() ?? ""
            let value = attributeDict["value"] ?? ""
            switch name {
            case "seeders":
                currentItem?.seeders = Int(value) ?? currentItem?.seeders ?? 0
            case "peers":
                currentItem?.peers = Int(value) ?? currentItem?.peers ?? 0
            case "size":
                currentItem?.size = Int64(value) ?? currentItem?.size ?? 0
            case "infohash":
                currentItem?.infoHash = value
            case "magneturl":
                currentItem?.magnetURL = value
            default:
                break
            }
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentText += string
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        let element = elementName.lowercased()
        guard var item = currentItem else { return }
        let value = currentText.trimmingCharacters(in: .whitespacesAndNewlines)

        switch element {
        case "title":
            item.title = value
        case "guid", "link":
            if value.lowercased().hasPrefix("magnet:?") {
                item.magnetURL = value
            }
        case "size":
            item.size = Int64(value) ?? item.size
        case "item":
            let title = item.title.isEmpty ? "Unknown" : item.title
            items.append(
                TorznabFeedItem(
                    title: title,
                    magnetURL: item.magnetURL,
                    infoHash: item.infoHash,
                    size: item.size,
                    seeders: item.seeders,
                    peers: item.peers
                )
            )
            currentItem = nil
        default:
            break
        }

        if element != "item" {
            currentItem = item
        }
    }

    private struct CurrentItem {
        var title = ""
        var magnetURL: String?
        var infoHash: String?
        var size: Int64 = 0
        var seeders: Int = 0
        var peers: Int = 0
    }
}

File: web/src/services/indexers/TorznabIndexer.ts
// Port of Sources/DebridStreamer/Services/Indexers/TorznabIndexer.swift.
//
// A generic Torznab/Jackett/Prowlarr XML-feed indexer. Mirrors the Swift actor:
// the same query params (t=search, imdbid/season/ep or q), category + apikey
// query handling, the X-Api-Key header mode for Prowlarr, non-2xx -> throw, and
// the Torznab feed parsing (title / magnet from guid|link|enclosure / size /
// torznab:attr seeders|peers|size|infohash|magneturl), with the infoHash
// fallback extracted from the magnet `xt=urn:btih:` parameter.
//
// Swift uses Foundation's event-driven `XMLParser`; there is no DOMParser in
// Node, so `parseTorznabFeed` below is a tiny hand-rolled tokenizer that
// reproduces the exact delegate event order/semantics.

import type { MediaType } from "../../models/media";
import { TorrentResult } from "./models";
import { defaultFetchImpl, type FetchImpl, IndexerError, type TorrentIndexer } from "./types";

export interface TorznabIndexerOptions {
  name: string;
  baseURL: string;
  endpointPath: string;
  apiKey?: string | null;
  categoryFilter?: string | null;
  sendAPIKeyAsHeader?: boolean;
  fetchImpl?: FetchImpl;
}

export class TorznabIndexer implements TorrentIndexer {
  readonly name: string;
  private readonly baseURL: string;
  private readonly endpointPath: string;
  private readonly apiKey: string | null;
  private readonly categoryFilter: string | null;
  private readonly sendAPIKeyAsHeader: boolean;
  private readonly fetchImpl: FetchImpl;

  constructor(options: TorznabIndexerOptions) {
    this.name = options.name;
    this.baseURL = options.baseURL;
    this.endpointPath = options.endpointPath;
    this.apiKey = options.apiKey ?? null;
    this.categoryFilter = options.categoryFilter ?? null;
    this.sendAPIKeyAsHeader = options.sendAPIKeyAsHeader ?? false;
    this.fetchImpl = options.fetchImpl ?? defaultFetchImpl;
  }

  async search(
    imdbId: string,
    _type: MediaType,
    season: number | null,
    episode: number | null,
  ): Promise<TorrentResult[]> {
    const params: Record<string, string> = {
      t: "search",
      imdbid: imdbId,
    };
    if (season != null) params.season = String(season);
    if (episode != null) params.ep = String(episode);
    return this.execute(params);
  }

  async searchByQuery(query: string, _type: MediaType): Promise<TorrentResult[]> {
    const params: Record<string, string> = {
      t: "search",
      q: query,
    };
    return this.execute(params);
  }

  private async execute(
    params: Record<string, string>,
  ): Promise<TorrentResult[]> {
    const { url, headers } = this.makeRequest(params);
    const response = await this.fetchImpl(url, headers ? { headers } : undefined);
    // Throw on non-2xx (consistent with the built-in indexers) so a
    // misconfigured endpoint / bad key surfaces as a recorded indexer failure
    // instead of an indistinguishable empty result. IndexerManager.searchAll
    // catches per-indexer, so this never aborts sibling indexers.
    if (!(response.status >= 200 && response.status <= 299)) {
      throw IndexerError.badServerResponse(response.status);
    }

    const items = parseTorznabFeed(await response.text());
    const results: TorrentResult[] = [];
    for (const item of items) {
      const hash = item.infoHash ?? extractInfoHash(item.magnetURL);
      if (hash == null || hash.length === 0) continue;
      results.push(
        TorrentResult.fromSearch({
          infoHash: hash,
          title: item.title,
          sizeBytes: item.size,
          seeders: item.seeders,
          leechers: item.peers,
          indexerName: this.name,
          magnetURI: item.magnetURL ?? null,
        }),
      );
    }
    return results;
  }

  /** Builds the request URL + optional header map. Mirrors Swift `makeRequest`,
   * incl. the `normalizedPath` join and the apikey-in-query vs X-Api-Key-header
   * split. Throws `IndexerError.badURL` when the base URL is unparseable. */
  private makeRequest(params: Record<string, string>): {
    url: string;
    headers: Record<string, string> | null;
  } {
    let url: URL;
    try {
      url = new URL(this.baseURL);
    } catch {
      throw IndexerError.badURL(this.baseURL);
    }

    if (this.endpointPath.length > 0) {
      url.pathname = normalizedPath(url.pathname, this.endpointPath);
    }

    for (const [k, v] of Object.entries(params)) {
      url.searchParams.append(k, v);
    }
    if (this.categoryFilter != null && this.categoryFilter.length > 0) {
      url.searchParams.append("cat", this.categoryFilter);
    }
    if (
      this.apiKey != null &&
      this.apiKey.length > 0 &&
      !this.sendAPIKeyAsHeader
    ) {
      url.searchParams.append("apikey", this.apiKey);
    }

    let headers: Record<string, string> | null = null;
    if (
      this.apiKey != null &&
      this.apiKey.length > 0 &&
      this.sendAPIKeyAsHeader
    ) {
      headers = { "X-Api-Key": this.apiKey };
    }

    return { url: url.toString(), headers };
  }
}

// MARK: - Path normalization (mirrors Swift `normalizedPath`)

/** Joins a base path with an endpoint, trimming slashes. Mirrors the Swift
 * `normalizedPath`. The DOM `URL` reports an empty path as "/", so the empty
 * base path is treated the same as Swift's empty string. */
function normalizedPath(currentPath: string, endpoint: string): string {
  const current = trimSlashes(currentPath);
  const append = trimSlashes(endpoint);
  if (current.length === 0) {
    return `/${append}`;
  }
  if (append.length === 0) {
    return `/${current}`;
  }
  return `/${current}/${append}`;
}

function trimSlashes(s: string): string {
  return s.replace(/^\/+/, "").replace(/\/+$/, "");
}

// MARK: - Magnet infoHash extraction (mirrors Swift `extractInfoHash`)

/** Pulls the btih hash out of a magnet URI's `xt=urn:btih:<hash>` param.
 * Mirrors Swift `extractInfoHash`. */
function extractInfoHash(magnetURL: string | null | undefined): string | null {
  if (magnetURL == null) return null;
  let components: URL;
  try {
    components = new URL(magnetURL);
  } catch {
    return null;
  }
  let xt: string | null = null;
  for (const [name, value] of components.searchParams) {
    if (name.toLowerCase() === "xt") {
      xt = value;
      break;
    }
  }
  if (xt == null) return null;
  const prefix = "urn:btih:";
  if (!xt.toLowerCase().startsWith(prefix)) return null;
  return xt.slice(prefix.length);
}

// MARK: - Torznab feed parser

interface TorznabFeedItem {
  title: string;
  magnetURL: string | null;
  infoHash: string | null;
  size: number;
  seeders: number;
  peers: number;
}

interface MutableItem {
  title: string;
  magnetURL: string | null;
  infoHash: string | null;
  size: number;
  seeders: number;
  peers: number;
}

/**
 * Parses a Torznab/RSS feed into items. Reproduces the exact event semantics of
 * the Swift `TorznabXMLDelegate`:
 *  - start `<item>` opens a fresh accumulator (and resets text);
 *  - `<enclosure url="...">` and `<torznab:attr name=.. value=..>` are read at
 *    start-element time (seeders/peers/size/infohash/magneturl);
 *  - text content is captured and, at end-element, applied to title /
 *    guid|link (magnet:? only) / size; `</item>` finalizes the item.
 */
export function parseTorznabFeed(xml: string): TorznabFeedItem[] {
  const items: TorznabFeedItem[] = [];
  let current: MutableItem | null = null;
  let currentText = "";

  const onStart = (
    name: string,
    attrs: Record<string, string>,
  ): void => {
    const element = name.toLowerCase();
    currentText = "";

    if (element === "item") {
      current = {
        title: "",
        magnetURL: null,
        infoHash: null,
        size: 0,
        seeders: 0,
        peers: 0,
      };
      return;
    }

    if (current == null) return;

    if (element === "enclosure" && attrs.url != null) {
      current.magnetURL = attrs.url;
    } else if (element === "torznab:attr") {
      const attrName = (attrs.name ?? "").toLowerCase();
      const value = attrs.value ?? "";
      switch (attrName) {
        case "seeders": {
          const n = Number.parseInt(value, 10);
          current.seeders = Number.isNaN(n) ? current.seeders : n;
          break;
        }
        case "peers": {
          const n = Number.parseInt(value, 10);
          current.peers = Number.isNaN(n) ? current.peers : n;
          break;
        }
        case "size": {
          const n = Number.parseInt(value, 10);
          current.size = Number.isNaN(n) ? current.size : n;
          break;
        }
        case "infohash":
          current.infoHash = value;
          break;
        case "magneturl":
          current.magnetURL = value;
          break;
        default:
          break;
      }
    }
  };

  const onEnd = (name: string): void => {
    const element = name.toLowerCase();
    if (current == null) return;
    const value = currentText.trim();

    switch (element) {
      case "title":
        current.title = value;
        break;
      case "guid":
      case "link":
        if (value.toLowerCase().startsWith("magnet:?")) {
          current.magnetURL = value;
        }
        break;
      case "size": {
        const n = Number.parseInt(value, 10);
        current.size = Number.isNaN(n) ? current.size : n;
        break;
      }
      case "item": {
        const title = current.title.length === 0 ? "Unknown" : current.title;
        items.push({
          title,
          magnetURL: current.magnetURL,
          infoHash: current.infoHash,
          size: current.size,
          seeders: current.seeders,
          peers: current.peers,
        });
        current = null;
        break;
      }
      default:
        break;
    }
  };

  tokenizeXML(xml, onStart, onEnd, (text) => {
    currentText += text;
  });

  return items;
}

// MARK: - Minimal streaming XML tokenizer

/**
 * A tiny SAX-style scanner: emits start-element (with parsed attributes),
 * end-element, and character events in document order. Handles self-closing
 * tags (emitting start then end), skips the XML declaration / comments / CDATA
 * boundaries, and decodes the common XML entities. This is deliberately small —
 * just enough to drive `parseTorznabFeed` the way Foundation's XMLParser drives
 * the Swift delegate.
 */
function tokenizeXML(
  xml: string,
  onStart: (name: string, attrs: Record<string, string>) => void,
  onEnd: (name: string) => void,
  onText: (text: string) => void,
): void {
  let i = 0;
  const n = xml.length;

  while (i < n) {
    const lt = xml.indexOf("<", i);
    if (lt === -1) {
      if (i < n) onText(decodeEntities(xml.slice(i)));
      break;
    }
    if (lt > i) {
      onText(decodeEntities(xml.slice(i, lt)));
    }

    // Declaration / processing instruction / comment / DOCTYPE / CDATA.
    if (xml.startsWith("<?", lt)) {
      const close = xml.indexOf("?>", lt);
      i = close === -1 ? n : close + 2;
      continue;
    }
    if (xml.startsWith("<!--", lt)) {
      const close = xml.indexOf("-->", lt);
      i = close === -1 ? n : close + 3;
      continue;
    }
    if (xml.startsWith("<![CDATA[", lt)) {
      const close = xml.indexOf("]]>", lt);
      const inner = xml.slice(lt + 9, close === -1 ? n : close);
      onText(inner); // CDATA is literal — no entity decoding.
      i = close === -1 ? n : close + 3;
      continue;
    }
    if (xml.startsWith("<!", lt)) {
      const close = xml.indexOf(">", lt);
      i = close === -1 ? n : close + 1;
      continue;
    }

    const gt = xml.indexOf(">", lt);
    if (gt === -1) {
      // Malformed trailing "<...": treat the rest as text.
      onText(decodeEntities(xml.slice(lt)));
      break;
    }

    let raw = xml.slice(lt + 1, gt);
    i = gt + 1;

    if (raw.startsWith("/")) {
      // End tag.
      const name = raw.slice(1).trim().split(/\s+/)[0];
      if (name.length > 0) onEnd(name);
      continue;
    }

    const selfClosing = raw.endsWith("/");
    if (selfClosing) raw = raw.slice(0, -1);

    const { name, attrs } = parseTag(raw);
    if (name.length === 0) continue;
    onStart(name, attrs);
    if (selfClosing) onEnd(name);
  }
}

/** Splits a tag body "name attr=\"v\" ..." into a name + attribute map. */
function parseTag(raw: string): { name: string; attrs: Record<string, string> } {
  const trimmed = raw.trim();
  const nameMatch = /^([^\s/>]+)/.exec(trimmed);
  if (nameMatch == null) return { name: "", attrs: {} };
  const name = nameMatch[1];
  const attrs: Record<string, string> = {};

  const attrRe = /([^\s=/]+)\s*=\s*("([^"]*)"|'([^']*)')/g;
  let m: RegExpExecArray | null;
  attrRe.lastIndex = nameMatch[0].length;
  while ((m = attrRe.exec(trimmed)) != null) {
    const key = m[1];
    const value = m[3] != null ? m[3] : (m[4] ?? "");
    attrs[key] = decodeEntities(value);
  }
  return { name, attrs };
}

/** Decodes the handful of XML entities that appear in feed text/attrs. */
function decodeEntities(s: string): string {
  if (!s.includes("&")) return s;
  return s
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&#(\d+);/g, (_, d: string) => String.fromCodePoint(Number(d)))
    .replace(/&#x([0-9a-fA-F]+);/g, (_, h: string) =>
      String.fromCodePoint(Number.parseInt(h, 16)),
    )
    .replace(/&amp;/g, "&");
}
